File: web/src/syncService.ts
// Chegga Web — incremental Chess.com ingestion into IndexedDB (Phase 1)
//
// Ported from Chegga's own `app/services/sync_service.py::sync_games`:
// safe to re-run at any time, upserts by chess_com_uuid so re-running
// never duplicates, and skips any month already marked "complete" except
// the current month, which can still receive new games.
//
// Two entry points share the same per-month logic (processMonth):
// - syncGames: the original full-history walk, oldest month first.
// - quickSyncRecentGames: added as a direct response to "the player might
//   click off before seeing what this does" -- walks newest month first
//   and stops once it's collected enough games, instead of always pulling
//   an entire account's history (94 months on the account this was
//   verified against) before a visitor sees anything. Months it does
//   fetch are marked "complete" exactly like the full sync would, so a
//   later syncGames() call for "get my whole history" correctly resumes
//   on whatever's left over -- no separate resume logic needed, this
//   reuses syncGames' own existing skip-already-complete behavior.

import { ChessComClient, ChessComHttpError } from "./chessComClient";
import type { ChessComRawGame } from "./chessComClient";
import { gameExists, getSyncState, putGame, putSyncState } from "./db";
import { normalizeGame, UntrackedUserError } from "./gameNormalizer";

export interface SyncProgress {
  monthsProcessed: number;
  totalMonths: number;
  gamesAdded: number;
  currentMonth?: string;
}

export interface SyncResult {
  monthsProcessed: number;
  gamesAdded: number;
}

function currentYearMonth(): string {
  const now = new Date();
  return `${now.getUTCFullYear()}-${String(now.getUTCMonth() + 1).padStart(2, "0")}`;
}

// https://api.chess.com/pub/player/{user}/games/{YYYY}/{MM}
function yearMonthFromArchiveUrl(url: string): string {
  const parts = url.replace(/\/+$/, "").split("/");
  const month = parts[parts.length - 1];
  const year = parts[parts.length - 2];
  return `${year}-${month}`;
}

/** Fetches and upserts one archive month, marking it "complete" in
 * syncState (unless it's the current, still-in-progress month). Returns
 * `undefined` if this month was already complete and should be skipped
 * (caller decides what "skipped" means for its own progress reporting).
 * Shared by both syncGames and quickSyncRecentGames so "what counts as
 * done" can never drift between the two. */
async function processMonth(
  db: IDBDatabase,
  client: ChessComClient,
  username: string,
  archiveUrl: string,
  currentMonth: string,
): Promise<{ yearMonth: string; addedThisMonth: number } | undefined> {
  const yearMonth = yearMonthFromArchiveUrl(archiveUrl);

  const state = await getSyncState(db, username, yearMonth);
  if (state && state.status === "complete" && yearMonth !== currentMonth) {
    return undefined;
  }

  let rawGames: ChessComRawGame[];
  try {
    rawGames = await client.getArchive(archiveUrl);
  } catch (err) {
    // A month Chess.com's own archive-list endpoint names can still 404 on
    // the archive endpoint itself (observed live against a real account,
    // not a documented API behavior) -- treat it as an empty month rather
    // than aborting the whole sync over one bad URL. Any other error
    // (network failure, a genuine server error after retries) still
    // propagates -- this is deliberately narrow.
    if (err instanceof ChessComHttpError && err.status === 404) {
      console.warn(`Chess.com archive 404 for ${yearMonth} (${archiveUrl}) -- treating as an empty month`);
      rawGames = [];
    } else {
      throw err;
    }
  }

  let addedThisMonth = 0;
  for (const raw of rawGames) {
    if (!raw.uuid) continue;
    if (await gameExists(db, raw.uuid)) continue;

    let record;
    try {
      record = normalizeGame(raw, username, username);
    } catch (err) {
      if (err instanceof UntrackedUserError) {
        console.warn(`Skipping game ${raw.uuid}: tracked user not found among players`);
        continue;
      }
      throw err;
    }

    await putGame(db, record);
    addedThisMonth += 1;
  }

  await putSyncState(db, {
    username,
    yearMonth,
    status: yearMonth !== currentMonth ? "complete" : "partial",
    gamesFetched: (state?.gamesFetched ?? 0) + addedThisMonth,
    lastSyncedAt: Date.now(),
  });

  return { yearMonth, addedThisMonth };
}

export async function syncGames(
  db: IDBDatabase,
  client: ChessComClient,
  username: string,
  onProgress?: (progress: SyncProgress) => void,
): Promise<SyncResult> {
  const archiveUrls = await client.getArchiveUrls(username);
  const currentMonth = currentYearMonth();
  const totalMonths = archiveUrls.length;

  let monthsProcessed = 0;
  let gamesAdded = 0;

  for (const archiveUrl of archiveUrls) {
    const result = await processMonth(db, client, username, archiveUrl, currentMonth);
    monthsProcessed += 1;
    if (result) gamesAdded += result.addedThisMonth;
    // Reported even on a skip (already-synced month) -- a skipped month is
    // still real progress through the account's history, just not a
    // fetch. Not reporting these made the progress bar look frozen for a
    // re-sync where most months are already done, then jump to "done" all
    // at once.
    onProgress?.({ monthsProcessed, totalMonths, gamesAdded, currentMonth: result?.yearMonth ?? yearMonthFromArchiveUrl(archiveUrl) });
  }

  return { monthsProcessed, gamesAdded };
}

/** Fast first-sync: walks archive months newest-first and stops once
 * `targetGameCount` new games have been collected (or the account's
 * whole history has been read, whichever comes first) -- so a visitor
 * with thousands of games gets a first payoff in however long it takes
 * to fetch a handful of recent months, not their entire history. Months
 * it does fetch are marked complete exactly like a full sync, so calling
 * syncGames() afterward ("get my whole history") correctly picks up only
 * what's left, oldest-first, with zero duplicate work. */
export async function quickSyncRecentGames(
  db: IDBDatabase,
  client: ChessComClient,
  username: string,
  targetGameCount: number,
  onProgress?: (progress: SyncProgress) => void,
): Promise<SyncResult & { fullyCaughtUp: boolean }> {
  const archiveUrls = await client.getArchiveUrls(username);
  const currentMonth = currentYearMonth();
  const totalMonths = archiveUrls.length;
  const newestFirst = [...archiveUrls].reverse();

  let monthsProcessed = 0;
  let gamesAdded = 0;

  for (const archiveUrl of newestFirst) {
    const result = await processMonth(db, client, username, archiveUrl, currentMonth);
    monthsProcessed += 1;
    if (result) gamesAdded += result.addedThisMonth;
    onProgress?.({ monthsProcessed, totalMonths, gamesAdded, currentMonth: result?.yearMonth ?? yearMonthFromArchiveUrl(archiveUrl) });

    if (gamesAdded >= targetGameCount) {
      return { monthsProcessed, gamesAdded, fullyCaughtUp: monthsProcessed >= totalMonths };
    }
  }

  return { monthsProcessed, gamesAdded, fullyCaughtUp: true };
}

File: archive/backend/app/services/sync_service.py
"""Incremental Chess.com ingestion -- safe to re-run at any time."""
import logging
from collections.abc import Callable
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.game import Game
from app.models.sync_state import SyncState
from app.services.chess_com_client import ChessComClient
from app.services.pgn_parser import normalize_game

logger = logging.getLogger(__name__)


def _current_year_month() -> str:
    now = datetime.now(timezone.utc)
    return f"{now.year:04d}-{now.month:02d}"


def _year_month_from_archive_url(url: str) -> str:
    # https://api.chess.com/pub/player/{user}/games/{YYYY}/{MM}
    parts = url.rstrip("/").split("/")
    year, month = parts[-2], parts[-1]
    return f"{year}-{month}"


def sync_games(
    db: Session,
    client: ChessComClient,
    username: str,
    on_progress: Callable[[int, int], None] | None = None,
) -> dict:
    """Pull every month of the user's Chess.com history not already marked
    complete. Games are upserted by chess_com_uuid, so re-running never
    duplicates rows; a month already fully fetched is skipped entirely
    except the current one, which can still receive new games.

    A first sync against a long history is many serial, throttled requests
    and can run for minutes -- on_progress(months_processed, games_added) is
    called after each month commits so a caller (the /api/sync/status route)
    can report real progress instead of looking stuck until the whole thing
    finishes."""
    archive_urls = client.get_archive_urls(username)
    current_month = _current_year_month()

    months_processed = 0
    games_added = 0

    for archive_url in archive_urls:
        year_month = _year_month_from_archive_url(archive_url)

        state = db.scalar(
            select(SyncState).where(SyncState.username == username, SyncState.year_month == year_month)
        )
        if state and state.status == "complete" and year_month != current_month:
            continue

        raw_games = client.get_archive(archive_url)
        added_this_month = 0

        for raw in raw_games:
            uuid = raw.get("uuid")
            if not uuid:
                continue
            exists = db.scalar(select(Game.id).where(Game.chess_com_uuid == uuid))
            if exists:
                continue
            try:
                fields = normalize_game(raw, username)
            except ValueError:
                logger.warning("Skipping game %s: tracked user not found among players", uuid)
                continue
            db.add(Game(**fields))
            added_this_month += 1

        if state is None:
            state = SyncState(username=username, year_month=year_month)
            db.add(state)
        state.status = "complete" if year_month != current_month else "partial"
        state.games_fetched = (state.games_fetched or 0) + added_this_month
        state.last_synced_at = datetime.utcnow()

        db.commit()

        months_processed += 1
        games_added += added_this_month
        logger.info("Synced %s: %d new games", year_month, added_this_month)
        if on_progress:
            on_progress(months_processed, games_added)

    return {"months_processed": months_processed, "games_added": games_added}
